✨ Collect the creation & update timestamps for KS sample workload

dumb0002 opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary

This PR provides scripts to collect the creation & update timestamp for the KS sample workload. Using these values, we'll be able to compute the e2e latency to down sync objects from a WDS to WEC and identify any bottleneck latency.

## Related issue(s)

Fixes #

MikeSpreitzer commented on test/performance/README.md:
```diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+
+   Below is a detailed explanation of the input parameters:
+   - `kubeconfig`: path to the kubeconfig file, e.g., *$HOME/.kube/config*
+   - `wds_ctx`: name of the context for the target WDS, e.g., *"wds1"*
```
In bash the dquotes around `wds1` are not needed; bash parsing rules will isolate the proper "word" ("word" is a technical term in bash syntax) with or without the dquotes. Also, squotes could be used instead of dquotes. In fact, there are many ways to write the same word.

- `wds1`
- `"wds1"`
- `'wds1'`
- `\wds1`
- `wd\s1`
- `"wds"1`
- `'wd'"s1"`
- and on and on

I think it is best to just write it in the plainest way possible, lest the reader mistakenly conclude that there is something special going on here.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/README.md:
```diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   - `its_ctx`: name of the context for the target ITS, e.g., *"its1"*
+   - `wec_ctx`: name of the context for the target WEC, e.g., *"cluster1"*
+   - `num_ns`: number of namespaces created in your experiment, e.g., *2*
+   - `output_dir`: path to the directory for the output data files, e.g., *$HOME/data/*
```
FYI, it would be nice to let the user know whether this has to be an absolute path.
FYI, it would be most ordinary if this does not have to be an absolute path.

dumb0002 replied:
It does not need to be an absolute path. The user just needs to provide the correct path to the directory. Fixed it to remove any confusion.

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   For each namespace, the creation and status_update timestamps will be collected for the workload objects created in `wds1`, manifestwork & workstatus objects created in `its1`, and workload & appliedmanifestworks objects created in `wec1`, for example: let's assume that `num_ns=2`, then the following directories and files will be created: 
+
+   ```bash 
+      $tree
````
What is `$tree`?

dumb0002 replied:
It is a recursive directory listing program that produces a depth-indented listing of files -- source: https://www.geeksforgeeks.org/tree-command-unixlinux/

MikeSpreitzer replied:
The dollar sign was part of the question.
It looks like you are missing a space between the dollar sign and the command name?

dumb0002 replied:
Fixed the space issue.

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   ```bash 
+      $tree
+      .
+      ├── perf-exp-0
````
Where are "perf-exp-0" and "perf-exp-1" coming from? The reader has not seen names like this before, has not been told what they correspond to. Are these the names of the namespaces generated for the experiment?

dumb0002 replied:
Yes, those are the names of the namespaces generated in the experiments. I provided an explanation in the section above: 
<img width="883" alt="Screenshot 2024-09-06 at 3 26 06 PM" src="https://github.com/user-attachments/assets/5680cdef-6d8a-42ad-87e7-a752a0038663">

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   The generated output file has the following structure:
+
+   ```bash
+   <obj_name> <obj_creation_time> <obj_status_update_time> <obj_status_condition>  <obj_status_condition>  <obj_controller_manager>
````
(A) Why is `<obj_status_condition>` listed twice on this line? It appears only once in the example below.

(B) What does "obj_controller_manager" mean? The example below is not illuminating.

dumb0002 replied:
(A) It was a typo - fixed it

(B) obj_controller_manager - it's the name of the controller that updates the status for the k8s object.

MikeSpreitzer replied:
Thank you for the explanation. I think it would be good to have an explanation in the document.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/metrics_collector.py:
```diff
@@ -98,29 +126,68 @@ def check_pod_status(self, namespace, watch_interval, label=None):
+   
+    # Long duration experiments parameters
+    numPods=2 # total number of workload objects set in your experiment
+    watch_interval = 60 # watch for events that to occur within time interval threshold
```
FYI, I would mention units here.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/metrics_collector.py:
```diff
@@ -98,29 +126,68 @@ def check_pod_status(self, namespace, watch_interval, label=None):
-
-    c = DataParser(numPods, kubeconfig, 'wds1', 'its1', 'wec1', f1, f2, f3, f4, f5)
-    c.check_pod_status(ns, watch_interval)
+    first = int(sys.argv[1])
+    second = int(sys.argv[2])
```
These two are not used for anything, right? Why are they here?

dumb0002 replied:
Oh! forgot to clean up those parameters - Done!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,58 @@ def load_kubeconfig(self, kubeconfig, context):
+            if name == None and ns != None and bindingPolicy != None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
+            elif  name != None and ns != None and bindingPolicy != None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
```
This is the same as line 62. Why two copies?

dumb0002 replied:
Not sure if I understood this comment correctly. There are different `if statements` that need to be handle in the same way

MikeSpreitzer replied:
Thanks for the explanation.

dumb0002 replied:
Fixed - consolidated `if statement`!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,58 @@ def load_kubeconfig(self, kubeconfig, context):
+            elif  name != None and ns != None and bindingPolicy != None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
+            elif name != None and ns != None and bindingPolicy == None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, field_selector='metadata.name=' + name)
```
Why use `list` instead of `get` here, since the name is known?

dumb0002 replied:
Fixed! using `get` instead to improve performance

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,58 @@ def load_kubeconfig(self, kubeconfig, context):
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
+            elif name != None and ns != None and bindingPolicy == None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, field_selector='metadata.name=' + name)
             else:
```
What if `name != None and ns == None` ?

dumb0002 replied:
The `else`  statement is already handling this scenario.

MikeSpreitzer replied:
Could this scenario be handled with a `get` instead of a `list`?

dumb0002 replied:
Add a new `if statement` to handle that case. Also added `check` flag in the condition to handle the case when we only know the name prefix for the workload object - as it is the case for the `appliedmanifestworks` (it starts with a randomly generated string and it does not include a label that match the corresponding binding policy as it is the case for the `manifestwork` objects)

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,58 @@ def load_kubeconfig(self, kubeconfig, context):
+
+            if name == None or single == False:
+               # Create the output file
+               if os.path.exists(fname):
+                  mode = 'a' # append if already exists
+               else:
+                  mode = 'w' # make a new file if not
```
Doesn't mode 'a' do the right thing in both cases?

```shell
mspreitz@mjs13 kubestellar % python3
Python 3.12.5 (main, Aug  6 2024, 19:08:49) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> fp = open('fubar', 'a')
>>> fp
<_io.TextIOWrapper name='fubar' mode='a' encoding='UTF-8'>
>>> fp.write('xyz')
3
>>> fp.close()
>>> ^D

mspreitz@mjs13 kubestellar % ls -l fubar
-rw-r--r--  1 mspreitz  staff  3 Sep  6 14:02 fubar
```

dumb0002 replied:
You're correct! I did similar test and I got the same results. Fixed!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,58 @@ def load_kubeconfig(self, kubeconfig, context):
                         status = object_list[i]["status"]
                     else:
-                        fout.write(name + "\t" + creation_time + "\t" + update_time + "\t" + status_condition + "\t" + op_manager + "\n")
+                        output=obj_name + "\t" + creation_timestamp + "\t" + update_timestamp + "\t" + status_condition + "\t" + op_manager
```
`op_manager == ""`, right? Is this what you intend?

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -106,72 +115,142 @@ def get_crd_time(self, grp, v, pl, name=None, check=False):
-            else: 
-                api_response = self.core_v1_api.list_namespaced_pod(ns)
+                if kind == "pod" or kind == "pods":
+                   api_response = self.core_v1_api.list_namespaced_pod(ns, field_selector='metadata.name=' + name)
```
Why use `list` instead of `get` when the name is known?

dumb0002 replied:
Fixed! using `get` instead to improve performance

MikeSpreitzer reviewed (commented):
I have finished a review. I left some individual comments. The only blocker I see is part (A) of https://github.com/kubestellar/kubestellar/pull/2416#pullrequestreview-2284573588 .

dumb0002 commented on test/performance/README.md:
````diff
@@ -111,13 +111,86 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   b) Run the metrics collection script:
+
+   ```bash 
+   $ python3 metrics_collector.py <kubeconfig> <wds_ctx> <its_ctx> <wec_ctx> <num_ns> <output_dir> 
````
@ezrasilvera , Fixed! I added the python3 in the requirement section at the top of the README. Also, update the changes of the input parameter of the collector script. 

@MikeSpreitzer, Fixed! Added a  `requirements.txt` file

dumb0002 commented:
> I have finished a review. I left some individual comments. The only blocker I see is part (A) of [#2416 (review)](https://github.com/kubestellar/kubestellar/pull/2416#pullrequestreview-2284573588) .

@MikeSpreitzer, thanks for the review. I addressed all your comments. This PR is ready for another look

MikeSpreitzer commented on test/performance/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+kubernetes==30.1.0
+PyYAML==5.4.1
+PyYAML==6.0.2
```
Two different versions for the same package looks wrong.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,54 @@ def load_kubeconfig(self, kubeconfig, context):
+            if name == None and ns != None and bindingPolicy != None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
+            elif  name != None and ns != None and bindingPolicy != None:
+               api_response = self.custom_objects_api.list_namespaced_custom_object(group=grp, version=v, plural=pl, namespace=ns, label_selector='transport.kubestellar.io/originOwnerReferenceBindingKey=' + bindingPolicy)
```
Why not use a `get` in this case?

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -51,48 +50,54 @@ def load_kubeconfig(self, kubeconfig, context):
+               # Create the output file
+               fout = open(fname, 'a')
 
             object_list = api_response["items"]
```
Is this right in the cases where `get` was used instead of `list`?

dumb0002 replied:
Yes, that's still correct. We still want to store the results in the output file.

MikeSpreitzer replied:
My question is whether `object_list = api_response["items"]` makes sense when `get` was invoked rather than `list`.

dumb0002 replied:
ok, I understood it now. I updated the script to better handle this scenario. Thanks!

MikeSpreitzer commented on test/performance/kshelper.py:
```diff
@@ -106,72 +111,138 @@ def get_crd_time(self, grp, v, pl, name=None, check=False):
+                # Create the output file
+                fout = open(fname, 'a')
 
             object_list = api_response.items
```
If this right if `name != None` ?

dumb0002 replied:
Yes, that's still correct. This condition has been tested.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+kubernetes==30.1.0
+PyYAML==5.4.1
+PyYAML==6.0.2
+urllib3==1.26.5
```
FYI, installing just kubernetes 30.1.0 brings in a lot of other stuff. Here is what I found:

```shell
(test5) mspreitz@mjs13 test5 % pip list
Package            Version
------------------ -----------
cachetools         5.5.0
certifi            2024.8.30
charset-normalizer 3.3.2
google-auth        2.34.0
idna               3.8
kubernetes         30.1.0
oauthlib           3.2.2
pip                24.2
pyasn1             0.6.0
pyasn1_modules     0.4.0
python-dateutil    2.9.0.post0
PyYAML             6.0.2
requests           2.32.3
requests-oauthlib  2.0.0
rsa                4.9
six                1.16.0
urllib3            2.2.2
websocket-client   1.8.0
```

Are you sure you want to downgrade urllib3?

dumb0002 replied:
Fixed the issue with `PyYAML` package - it was a typo.  The command `pip list` shows all the package installed in your system. Are you sure that all those packages didn't already exist before you installed the `Kubernetes` module?  I see a long list of package on my testing machine even without installing the `Kubernetes` module first.

MikeSpreitzer replied:
Yes, I see only `pip` itself before installing kubernetes.

```
mspreitz@mjs13 ~ % python3 -m venv test5
mspreitz@mjs13 ~ % . test5/bin/activate
(test5) mspreitz@mjs13 ~ % pip list
Package Version
------- -------
pip     24.2
(test5) mspreitz@mjs13 ~ % pip install kubernetes==30.1.0
Collecting kubernetes==30.1.0
  Using cached kubernetes-30.1.0-py2.py3-none-any.whl.metadata (1.5 kB)
Collecting certifi>=14.05.14 (from kubernetes==30.1.0)
  Using cached certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
Collecting six>=1.9.0 (from kubernetes==30.1.0)
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting python-dateutil>=2.5.3 (from kubernetes==30.1.0)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting pyyaml>=5.4.1 (from kubernetes==30.1.0)
  Using cached PyYAML-6.0.2-cp312-cp312-macosx_11_0_arm64.whl.metadata (2.1 kB)
Collecting google-auth>=1.0.1 (from kubernetes==30.1.0)
  Using cached google_auth-2.34.0-py2.py3-none-any.whl.metadata (4.7 kB)
Collecting websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0 (from kubernetes==30.1.0)
  Using cached websocket_client-1.8.0-py3-none-any.whl.metadata (8.0 kB)
Collecting requests (from kubernetes==30.1.0)
  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting requests-oauthlib (from kubernetes==30.1.0)
  Using cached requests_oauthlib-2.0.0-py2.py3-none-any.whl.metadata (11 kB)
Collecting oauthlib>=3.2.2 (from kubernetes==30.1.0)
  Using cached oauthlib-3.2.2-py3-none-any.whl.metadata (7.5 kB)
Collecting urllib3>=1.24.2 (from kubernetes==30.1.0)
  Using cached urllib3-2.2.2-py3-none-any.whl.metadata (6.4 kB)
Collecting cachetools<6.0,>=2.0.0 (from google-auth>=1.0.1->kubernetes==30.1.0)
  Using cached cachetools-5.5.0-py3-none-any.whl.metadata (5.3 kB)
Collecting pyasn1-modules>=0.2.1 (from google-auth>=1.0.1->kubernetes==30.1.0)
  Using cached pyasn1_modules-0.4.0-py3-none-any.whl.metadata (3.4 kB)
Collecting rsa<5,>=3.1.4 (from google-auth>=1.0.1->kubernetes==30.1.0)
  Using cached rsa-4.9-py3-none-any.whl.metadata (4.2 kB)
Collecting charset-normalizer<4,>=2 (from requests->kubernetes==30.1.0)
  Using cached charset_normalizer-3.3.2-cp312-cp312-macosx_11_0_arm64.whl.metadata (33 kB)
Collecting idna<4,>=2.5 (from requests->kubernetes==30.1.0)
  Using cached idna-3.8-py3-none-any.whl.metadata (9.9 kB)
Collecting pyasn1<0.7.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth>=1.0.1->kubernetes==30.1.0)
  Using cached pyasn1-0.6.0-py2.py3-none-any.whl.metadata (8.3 kB)
Using cached kubernetes-30.1.0-py2.py3-none-any.whl (1.7 MB)
Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)
Using cached google_auth-2.34.0-py2.py3-none-any.whl (200 kB)
Using cached oauthlib-3.2.2-py3-none-any.whl (151 kB)
Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Using cached PyYAML-6.0.2-cp312-cp312-macosx_11_0_arm64.whl (173 kB)
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Using cached urllib3-2.2.2-py3-none-any.whl (121 kB)
Using cached websocket_client-1.8.0-py3-none-any.whl (58 kB)
Using cached requests-2.32.3-py3-none-any.whl (64 kB)
Using cached requests_oauthlib-2.0.0-py2.py3-none-any.whl (24 kB)
Using cached cachetools-5.5.0-py3-none-any.whl (9.5 kB)
Using cached charset_normalizer-3.3.2-cp312-cp312-macosx_11_0_arm64.whl (119 kB)
Using cached idna-3.8-py3-none-any.whl (66 kB)
Using cached pyasn1_modules-0.4.0-py3-none-any.whl (181 kB)
Using cached rsa-4.9-py3-none-any.whl (34 kB)
Using cached pyasn1-0.6.0-py2.py3-none-any.whl (85 kB)
Installing collected packages: websocket-client, urllib3, six, pyyaml, pyasn1, oauthlib, idna, charset-normalizer, certifi, cachetools, rsa, requests, python-dateutil, pyasn1-modules, requests-oauthlib, google-auth, kubernetes
Successfully installed cachetools-5.5.0 certifi-2024.8.30 charset-normalizer-3.3.2 google-auth-2.34.0 idna-3.8 kubernetes-30.1.0 oauthlib-3.2.2 pyasn1-0.6.0 pyasn1-modules-0.4.0 python-dateutil-2.9.0.post0 pyyaml-6.0.2 requests-2.32.3 requests-oauthlib-2.0.0 rsa-4.9 six-1.16.0 urllib3-2.2.2 websocket-client-1.8.0
(test5) mspreitz@mjs13 ~ % 
(test5) mspreitz@mjs13 ~ % pip list
Package            Version
------------------ -----------
cachetools         5.5.0
certifi            2024.8.30
charset-normalizer 3.3.2
google-auth        2.34.0
idna               3.8
kubernetes         30.1.0
oauthlib           3.2.2
pip                24.2
pyasn1             0.6.0
pyasn1_modules     0.4.0
python-dateutil    2.9.0.post0
PyYAML             6.0.2
requests           2.32.3
requests-oauthlib  2.0.0
rsa                4.9
six                1.16.0
urllib3            2.2.2
websocket-client   1.8.0
(test5) mspreitz@mjs13 ~ % 
```

dumb0002 replied:
ok, I see! I updated the `requirements.txt` file to only include the Kubernetes module

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,88 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   <obj-name> <obj-creation-time> <obj-status-update-time> <obj-status-condition> <obj-controller-manager>
+   ```
+   
+   Where: `<obj-controller-manager>` - it is the name of the k8s controller that updates the status of the workload object.
````
FYI, I would be a little more precise, something like "the FieldManager value used by the controller that updated...".

MikeSpreitzer commented on test/performance/README.md:
```diff
@@ -1,6 +1,6 @@
+*Pre-requisite*: in order to follow the instructions below, you must have [python3](https://www.python.org/downloads/) and a environment with KubeStellar installed; see [KubeStellar getting started](https://docs.kubestellar.io/release-0.23.1/direct/get-started/). Alternatively, you can also use KubeStellar e2e script [run-test.sh](https://github.com/kubestellar/kubestellar/blob/main/test/e2e/run- ...
 
 To generate the sample workload for KubeStellar performance experiments proceed as following:
 
```
I tried testing, and ran into trouble pretty quickly. Here is the tail of it:

```shell
mspreitz@mjs13 kubestellar % cd test/performance 
mspreitz@mjs13 performance % ./setup-clusterloader2.sh 
+ set -e
+ env=kind
+ '[' '' == --env ']'
+ :
+ : -------------------------------------------------------------------------
+ : 'Configuring clusterloader2 to generate KS performance workloads'
+++ dirname -- ./setup-clusterloader2.sh
++ cd -- .
++ pwd
+ SCRIPT_DIR=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test/performance
+ cp /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test/performance/cleanup.sh /Users/mspreitz/go/src/k8s.io/perf-tests/
+ cp /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test/performance/workloads/modules/cluster-density-k8s.yaml /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test/performance/workloads/modules/cluster-density-ocp.yaml /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/test/performance/workloads/modules/performance-test-module.yaml /Users/mspreitz/go/src/k8s.io/perf-tests/testing/load/modules
cp: /Users/mspreitz/go/src/k8s.io/perf-tests/testing/load/modules is not a directory
mspreitz@mjs13 performance % 
mspreitz@mjs13 performance % echo $?
1
mspreitz@mjs13 performance % ls $CL2_DIR
CONTRIBUTING.md				_hook					network
LICENSE					adhoc					perfdash
Makefile				application-benchmarks			run-e2e-with-prometheus-fw-rule.sh
OWNERS					benchmark				run-e2e.sh
OWNERS_ALIASES				cleanup.sh				slo-monitor
README.md				clusterloader2				util-images
RELEASE.md				code-of-conduct.md			verify
SECURITY.md				dns
SECURITY_CONTACTS			golang
mspreitz@mjs13 performance % 
```

MikeSpreitzer replied:
https://github.com/kubernetes/perf-tests/tree/release-1.31 shows no directory named "testing".

dumb0002 replied:
@MikeSpreitzer, it seems that you didn't set your CL2_DIR correctly. You should set it as following: 
```
export CL2_DIR=$HOME/perf-tests/clusterloader2
```

dumb0002 replied:
this command `ls $CL2_DIR` should show the contents of this directory instead:  `path/perf-tests/clusterloader2`

MikeSpreitzer replied:
With a little `mkdir -p`, I was then able to invoke that script.

Later, I got another failure:

```shell
mspreitz@mjs13 perf-tests % cd $CL2_DIR
mspreitz@mjs13 perf-tests % go run cmd/clusterloader.go --testconfig=./testing/load/performance-test-config.yaml --kubeconfig=$HOME/.kube/config --provider=ks --v=2
stat cmd/clusterloader.go: no such file or directory
mspreitz@mjs13 perf-tests % ls cmd
ls: cmd: No such file or directory
```

MikeSpreitzer commented on test/performance/README.md:
```diff
@@ -111,13 +111,88 @@ To generate the sample workload for KubeStellar performance experiments proceed
     - 10 config maps containing a 2048-character random string
 
 
```
I tried testing with a plain Kube cluster made with `test/e2e/run-test.sh` in this PR. The run of clusterloader produced some error messages (lines starting with `E`) but terminated with status 0. I am not sure whether there was a problem or not.

```shell
mspreitz@mjs13 clusterloader2 % go run cmd/clusterloader.go --testconfig=./testing/load/performance-test-config.yaml --kubeconfig=$HOME/.kube/config --provider=ks --v=2
I0909 16:13:25.937222   26877 dns_performance-k8s-hostnames.go:58] Registering measurement: DNS Performance for K8s Hostnames
I0909 16:13:25.937339   26877 network_performance_measurement.go:90] Registering Network Performance Measurement
I0909 16:13:25.939261   26877 clusterloader.go:267] Listening on 8000
I0909 16:13:25.956524   26877 clusterloader.go:163] ClusterConfig.Nodes set to 0
E0909 16:13:25.960706   26877 clusterloader.go:171] Getting master name error: master node not found
E0909 16:13:25.964532   26877 clusterloader.go:180] Getting master external ip error: didn't find any ExternalIP master IPs
E0909 16:13:25.969165   26877 clusterloader.go:189] Getting master internal ip error: didn't find any InternalIP master IPs
I0909 16:13:25.969184   26877 clusterloader.go:295] Using config: {ClusterConfig:{KubeConfigPath:/Users/mspreitz/.kube/config RunFromCluster:false Nodes:0 Provider:0x14000704e90 EtcdCertificatePath:/etc/srv/kubernetes/pki/etcd-apiserver-server.crt EtcdKeyPath:/etc/srv/kubernetes/pki/etcd-apiserver-server.key EtcdInsecurePort:2382 MasterIPs:[] MasterInternalIPs:[] MasterName: DeleteStaleNamespaces:false DeleteAutomanagedNamespaces:true APIServerPprofByClientEnabled:false KubeletPort:10250 K8SClientsNumber:1 SkipClusterVerification:false} ReportDir: ExecServiceConfig:{Enable:false} ModifierConfig:{OverwriteTestConfig:[] SkipSteps:[]} PrometheusConfig:{TearDownServer:true EnableServer:false EnablePushgateway:false ScrapeEtcd:false ScrapeNodeExporter:false ScrapeWindowsNodeExporter:false ScrapeKubelets:false ScrapeMasterKubelets:false ScrapeKubeProxy:true KubeProxySelectorKey:component ScrapeKubeStateMetrics:false ScrapeMetricsServerMetrics:false ScrapeNodeLocalDNS:false ScrapeAnet:false ScrapeNetworkPolicies:false ScrapeMastersWithPublicIPs:false APIServerScrapePort:443 SnapshotProject: AdditionalMonitorsPath: StorageClassProvisioner:kubernetes.io/gce-pd StorageClassVolumeType:pd-ssd PVCStorageClass:ssd ReadyTimeout:15m0s PrometheusMemoryRequest:10Gi} OverridePaths:[]}
I0909 16:13:25.974224   26877 cluster.go:75] Listing cluster nodes:
I0909 16:13:25.974286   26877 framework.go:74] Creating framework with 1 clients and "/Users/mspreitz/.kube/config" kubeconfig.
W0909 16:13:25.981268   26877 imagepreload.go:92] No images specified. Skipping image preloading
I0909 16:13:25.985203   26877 clusterloader.go:449] Test config successfully dumped to: generatedConfig_load.yaml
I0909 16:13:25.985215   26877 clusterloader.go:238] --------------------------------------------------------------------------------
I0909 16:13:25.985221   26877 clusterloader.go:239] Running ./testing/load/performance-test-config.yaml
I0909 16:13:25.985227   26877 clusterloader.go:240] --------------------------------------------------------------------------------
I0909 16:13:25.985266   26877 simple_test_executor.go:58] AutomanagedNamespacePrefix: test-1itr3m
I0909 16:13:26.005370   26877 simple_test_executor.go:149] Step "[step: 01] Creating Namespaces and BindingPolicies Objects" started
I0909 16:13:26.037831   26877 simple_test_executor.go:171] Step "[step: 01] Creating Namespaces and BindingPolicies Objects" ended
I0909 16:13:26.037880   26877 simple_test_executor.go:149] Step "[step: 02] Wait 10s" started
I0909 16:13:36.049647   26877 simple_test_executor.go:171] Step "[step: 02] Wait 10s" ended
I0909 16:13:36.049878   26877 simple_test_executor.go:149] Step "[step: 03] Creating Workload k8s Objects for Namespace: perf-exp-0" started
I0909 16:13:37.151471   26877 simple_test_executor.go:171] Step "[step: 03] Creating Workload k8s Objects for Namespace: perf-exp-0" ended
I0909 16:13:37.151547   26877 simple_test_executor.go:149] Step "[step: 04] Creating Workload k8s Objects for Namespace: perf-exp-1" started
I0909 16:13:38.294247   26877 simple_test_executor.go:171] Step "[step: 04] Creating Workload k8s Objects for Namespace: perf-exp-1" ended
I0909 16:13:38.294370   26877 simple_test_executor.go:80] Waiting for the chaos monkey subroutine to end...
I0909 16:13:38.294412   26877 simple_test_executor.go:82] Chaos monkey ended.
I0909 16:13:38.294431   26877 simple_test_executor.go:102] 
I0909 16:13:48.310682   26877 simple_test_executor.go:400] Resources cleanup time: 10.01622225s
I0909 16:13:48.310758   26877 clusterloader.go:248] --------------------------------------------------------------------------------
I0909 16:13:48.310777   26877 clusterloader.go:249] Test Finished
I0909 16:13:48.310798   26877 clusterloader.go:250]   Test: ./testing/load/performance-test-config.yaml
I0909 16:13:48.310810   26877 clusterloader.go:251]   Status: Success
I0909 16:13:48.310820   26877 clusterloader.go:255] --------------------------------------------------------------------------------

JUnit report was created: /Users/mspreitz/go/src/k8s.io/perf-tests/clusterloader2/junit.xml
I0909 16:13:48.311539   26877 prometheus.go:307] Get snapshot from Prometheus
mspreitz@mjs13 clusterloader2 % echo $?
0
```

dumb0002 replied:
@MikeSpreitzer, Great! that's a successful run of the clusterloader. You can ignore the error message shown above because were creating workloads in a WDS space instead of a vanilla k8s environment.

MikeSpreitzer replied:
It would be best if the script did not print error messages when there is not an error. Second best would be to document when the error messages can safely be ignored.

dumb0002 replied:
That's the normal output of the clusterloader2 open source script. From my point of view the error message is self-explanatory: there isn't any master node with external IP address in a WDS space.

dumb0002 replied:
The final output of the script clearly shows that the test was successful: 

```
I0909 16:13:48.310758   26877 clusterloader.go:248] --------------------------------------------------------------------------------
I0909 16:13:48.310777   26877 clusterloader.go:249] Test Finished
I0909 16:13:48.310798   26877 clusterloader.go:250]   Test: ./testing/load/performance-test-config.yaml
I0909 16:13:48.310810   26877 clusterloader.go:251]   Status: Success
I0909 16:13:48.310820   26877 clusterloader.go:255] --------------------------------------------------------------------------------

```

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,88 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   ```bash
+   $ cd .$HOME/kubestellar/test/performance
+   ```
+
````
At this point it would be worth pointing out that the following depends on the dependencies listed in `requirements.txt`.

dumb0002 replied:
Done!

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,88 @@ To generate the sample workload for KubeStellar performance experiments proceed
+
+   ```bash 
+      $ tree
+      .
````
A normal Python user would have created a virtual environment named `.venv` at this point and installed the dependencies in it. That will lead this listing to be full of a huge amount of stuff besides experiment results.

dumb0002 replied:
Indeed. That's a user preference of how to set-up his Python env and install the required requirements. From my point of view those instructions is out-of-scope from this README.

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -111,13 +111,88 @@ To generate the sample workload for KubeStellar performance experiments proceed
+   For example:
+   
+   ```bash 
+   $ python3 metrics_collector.py $HOME/.kube/config wds1 its1 cluster1 2 $HOME/data
````
I tested this. The script ran without producing any error messages, and it also produced no output files. Was I supposed to run it before the cleanup script? If so then there is a problem with the narrative order in these instructions.

MikeSpreitzer replied:
I tested again without running the cleanup script. Still no output file. BTW, should I look for them in `.` or `<output-directory>` ? If the latter, then the example should be adjusted. Even so, I found no output in either place.

```shell
(.venv) bash-3.2$ ls $HOME/data
(.venv) bash-3.2$ python3 metrics_collector.py $HOME/.kube/config wds1 its1 cluster1 2 $HOME/data
(.venv) bash-3.2$ find $HOME/data
/Users/mspreitz/data
(.venv) bash-3.2$ ls
README.md		cleanup.sh		metrics_collector.py	setup-clusterloader2.sh
__pycache__		kshelper.py		requirements.txt	workloads
(.venv) bash-3.2$ kubectl get ns
NAME                 STATUS   AGE
kube-system          Active   83m
kube-public          Active   83m
kube-node-lease      Active   83m
default              Active   83m
kubestellar-report   Active   82m
nginx                Active   81m
perf-exp-1           Active   2m16s
perf-exp-0           Active   2m16s
(.venv) bash-3.2$ 
(.venv) bash-3.2$ find . -name "perf-exp*"
(.venv) bash-3.2$ 
```

dumb0002 replied:
@MikeSpreitzer, I just pushed a patch for that, could you re-run again the following command?

```
$ python3 metrics_collector.py $HOME/.kube/config wds1 its1 cluster1 2 $HOME/data
```

Yes, you should run it without running the cleanup script. I will fix the README to address that now.

MikeSpreitzer commented on test/performance/README.md:
```diff
@@ -1,6 +1,6 @@
+*Pre-requisite*: in order to follow the instructions below, you must have [python3](https://www.python.org/downloads/) and a environment with KubeStellar installed; see [KubeStellar getting started](https://docs.kubestellar.io/release-0.23.1/direct/get-started/). Alternatively, you can also use KubeStellar e2e script [run-test.sh](https://github.com/kubestellar/kubestellar/blob/main/test/e2e/run- ...
 
 To generate the sample workload for KubeStellar performance experiments proceed as following:
 
```
Yes, that's why I deleted my original comment.

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -112,12 +112,89 @@ To generate the sample workload for KubeStellar performance experiments proceed
+
+   ```bash 
+      $ tree
+      .
````
The output goes into the output directory, right? In none of the preceding examples is `.` the output directory. I would stick with `$HOME/data`, that seems to be the more commonly used example in the earlier text.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on test/performance/README.md:
````diff
@@ -112,12 +112,89 @@ To generate the sample workload for KubeStellar performance experiments proceed
+         └── workstatuses
+   ```
+
+   The generated output file has the following structure:
````
More than one file is generated.

MikeSpreitzer commented:
I tested the latest version of this PR, commit 5c0b0fa6e1cf1ab86c84c36b2ded70b4bb7b2f6f, and this time got some data file. But most were empty. Is this what I should expect? I used a plain kube cluster created by `test/e2e/run-test.sh`.

```shell
(.venv) bash-3.2$ wc $(find $HOME/data -type f)
       0       0       0 /Users/mspreitz/data/perf-exp-0/secrets-wec
       0       0       0 /Users/mspreitz/data/perf-exp-0/configmaps-wec
      11      33     483 /Users/mspreitz/data/perf-exp-0/configmaps-wds1
       0       0       0 /Users/mspreitz/data/perf-exp-0/services-wec
       1       4      56 /Users/mspreitz/data/perf-exp-0/deployments-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-0/services-wds1
       2       8     313 /Users/mspreitz/data/perf-exp-0/appliedmanifestworks
       0       0       0 /Users/mspreitz/data/perf-exp-0/workstatuses
       0       0       0 /Users/mspreitz/data/perf-exp-0/manifestworks
       0       0       0 /Users/mspreitz/data/perf-exp-0/deployments-wec
      10      30     440 /Users/mspreitz/data/perf-exp-0/secrets-wds1
       0       0       0 /Users/mspreitz/data/perf-exp-1/secrets-wec
       0       0       0 /Users/mspreitz/data/perf-exp-1/configmaps-wec
      11      33     483 /Users/mspreitz/data/perf-exp-1/configmaps-wds1
       0       0       0 /Users/mspreitz/data/perf-exp-1/services-wec
       1       4      56 /Users/mspreitz/data/perf-exp-1/deployments-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-1/services-wds1
       2       8     313 /Users/mspreitz/data/perf-exp-1/appliedmanifestworks
       0       0       0 /Users/mspreitz/data/perf-exp-1/workstatuses
       0       0       0 /Users/mspreitz/data/perf-exp-1/manifestworks
       0       0       0 /Users/mspreitz/data/perf-exp-1/deployments-wec
      10      30     440 /Users/mspreitz/data/perf-exp-1/secrets-wds1
      54     168    2866 total
```

MikeSpreitzer commented:
Also, the contents were less than promised.

```shell
(.venv) bash-3.2$ grep -n . ~/data/perf-exp-0/configmaps-wds1
1:kube-root-ca.crt	2024-09-10 05:20:28+00:00
2:cluster-density-0	2024-09-10 05:20:38+00:00
3:cluster-density-1	2024-09-10 05:20:38+00:00
4:cluster-density-2	2024-09-10 05:20:38+00:00
5:cluster-density-3	2024-09-10 05:20:38+00:00
6:cluster-density-4	2024-09-10 05:20:38+00:00
7:cluster-density-5	2024-09-10 05:20:38+00:00
8:cluster-density-6	2024-09-10 05:20:38+00:00
9:cluster-density-7	2024-09-10 05:20:38+00:00
10:cluster-density-8	2024-09-10 05:20:39+00:00
11:cluster-density-9	2024-09-10 05:20:39+00:00
```

MikeSpreitzer commented:
It looks like the test workload did not go to the WECs.

```
(.venv) bash-3.2$ kubectl --context cluster1 get ns
NAME                                  STATUS   AGE
default                               Active   18m
kube-node-lease                       Active   18m
kube-public                           Active   18m
kube-system                           Active   18m
local-path-storage                    Active   18m
nginx                                 Active   17m
open-cluster-management               Active   18m
open-cluster-management-agent         Active   18m
open-cluster-management-agent-addon   Active   17m
(.venv) bash-3.2$ 
```

MikeSpreitzer commented:
```shell
(.venv) bash-3.2$ kubectl --context wds1 get Binding.control perf-exp-0 -o yaml
apiVersion: control.kubestellar.io/v1alpha1
kind: Binding
metadata:
  creationTimestamp: "2024-09-10T05:20:30Z"
  finalizers:
  - transport.kubestellar.io/object-cleanup
  generation: 1
  name: perf-exp-0
  ownerReferences:
  - apiVersion: control.kubestellar.io/v1alpha1
    blockOwnerDeletion: false
    controller: true
    kind: BindingPolicy
    name: perf-exp-0
    uid: 75b39ff2-02ac-45fb-9079-e20e474ca209
  resourceVersion: "490"
  uid: ec7d58a3-4749-4731-9b51-f1225d011116
spec:
  destinations:
  - clusterId: cluster1
  workload: {}
status:
  observedGeneration: 1
```

MikeSpreitzer commented:
```shell
(.venv) bash-3.2$ kubectl --context wds1 get BindingPolicy.control perf-exp-0 -o yaml
apiVersion: control.kubestellar.io/v1alpha1
kind: BindingPolicy
metadata:
  creationTimestamp: "2024-09-10T05:20:28Z"
  finalizers:
  - bindingpolicy.kubestellar.io/kscontroller
  generation: 1
  labels:
    kubestellar: performance-test
  name: perf-exp-0
  resourceVersion: "476"
  uid: 75b39ff2-02ac-45fb-9079-e20e474ca209
spec:
  clusterSelectors:
  - matchLabels:
      location-group: edge
  downsync:
  - objectSelectors:
    - matchLabels:
        app.kubernetes.io/name: cluster-density-"0"
  wantSingletonReportedState: true
```

MikeSpreitzer commented:
```shell
(.venv) bash-3.2$ kubectl --context wds1 get ns perf-exp-0 -o yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2024-09-10T05:20:28Z"
  labels:
    app.kubernetes.io/name: cluster-density-0
    kubernetes.io/metadata.name: perf-exp-0
    kubestellar: performance-test
  name: perf-exp-0
  resourceVersion: "477"
  uid: a246d430-5295-416e-8950-d9daf7faf1d8
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

dumb0002 commented:
@MikeSpreitzer, I added the fix for the `bindingPolicy` quotes issue - you should be able to see now the workload objects being synced to the target WEC as expected.

MikeSpreitzer commented:
I am testing the latest version. Commit 754e4a9a67b655a175db06f7cb0fc475cf8d8977.

I noticed that the end of the output from running clusterlloader2 is

```shell
JUnit report was created: /Users/mspreitz/go/src/k8s.io/perf-tests/clusterloader2/junit.xml
I0910 13:34:16.850798    5974 prometheus.go:307] Get snapshot from Prometheus
```

And the exit status is 0. Where is the Prometheus snapshot? BTW, as far as I know, my system does not have Prometheus installed.

MikeSpreitzer commented:
This time I found more content in the output files:

```shell
(.venv) bash-3.2$ wc $(find $HOME/data -type f)
      10      30     440 /Users/mspreitz/data/perf-exp-0/secrets-wec
      11      33     483 /Users/mspreitz/data/perf-exp-0/configmaps-wec
      11      33     483 /Users/mspreitz/data/perf-exp-0/configmaps-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-0/services-wec
       1       7      99 /Users/mspreitz/data/perf-exp-0/deployments-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-0/services-wds1
       1       4     146 /Users/mspreitz/data/perf-exp-0/appliedmanifestworks
       1       4     142 /Users/mspreitz/data/perf-exp-0/workstatuses
       1       4      80 /Users/mspreitz/data/perf-exp-0/manifestworks
       1       7     104 /Users/mspreitz/data/perf-exp-0/deployments-wec
      10      30     440 /Users/mspreitz/data/perf-exp-0/secrets-wds1
      10      30     440 /Users/mspreitz/data/perf-exp-1/secrets-wec
      11      33     483 /Users/mspreitz/data/perf-exp-1/configmaps-wec
      11      33     483 /Users/mspreitz/data/perf-exp-1/configmaps-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-1/services-wec
       1       7      99 /Users/mspreitz/data/perf-exp-1/deployments-wds1
       3       9     141 /Users/mspreitz/data/perf-exp-1/services-wds1
       1       4     146 /Users/mspreitz/data/perf-exp-1/appliedmanifestworks
       1       4     142 /Users/mspreitz/data/perf-exp-1/workstatuses
       1       4      80 /Users/mspreitz/data/perf-exp-1/manifestworks
       1       7     104 /Users/mspreitz/data/perf-exp-1/deployments-wec
      10      30     440 /Users/mspreitz/data/perf-exp-1/secrets-wds1
     106     340    5398 total
```

Some example contents (with .txt appended to the filename):

[deployments-wds1.txt](https://github.com/user-attachments/files/16948850/deployments-wds1.txt)
[deployments-wec.txt](https://github.com/user-attachments/files/16948851/deployments-wec.txt)
[services-wds1.txt](https://github.com/user-attachments/files/16948853/services-wds1.txt)
[services-wec.txt](https://github.com/user-attachments/files/16948854/services-wec.txt)

MikeSpreitzer commented:
The output files are CSV files using tab as the delimiter. It might be good for their names to have the `.csv` extension.

dumb0002 commented:
> The output files are CSV files using tab as the delimiter. It might be good for their names to have the `.csv` extension.

Done!

dumb0002 commented:
> This time I found more content in the output files:
> 
> ```shell
> (.venv) bash-3.2$ wc $(find $HOME/data -type f)
>       10      30     440 /Users/mspreitz/data/perf-exp-0/secrets-wec
>       11      33     483 /Users/mspreitz/data/perf-exp-0/configmaps-wec
>       11      33     483 /Users/mspreitz/data/perf-exp-0/configmaps-wds1
>        3       9     141 /Users/mspreitz/data/perf-exp-0/services-wec
>        1       7      99 /Users/mspreitz/data/perf-exp-0/deployments-wds1
>        3       9     141 /Users/mspreitz/data/perf-exp-0/services-wds1
>        1       4     146 /Users/mspreitz/data/perf-exp-0/appliedmanifestworks
>        1       4     142 /Users/mspreitz/data/perf-exp-0/workstatuses
>        1       4      80 /Users/mspreitz/data/perf-exp-0/manifestworks
>        1       7     104 /Users/mspreitz/data/perf-exp-0/deployments-wec
>       10      30     440 /Users/mspreitz/data/perf-exp-0/secrets-wds1
>       10      30     440 /Users/mspreitz/data/perf-exp-1/secrets-wec
>       11      33     483 /Users/mspreitz/data/perf-exp-1/configmaps-wec
>       11      33     483 /Users/mspreitz/data/perf-exp-1/configmaps-wds1
>        3       9     141 /Users/mspreitz/data/perf-exp-1/services-wec
>        1       7      99 /Users/mspreitz/data/perf-exp-1/deployments-wds1
>        3       9     141 /Users/mspreitz/data/perf-exp-1/services-wds1
>        1       4     146 /Users/mspreitz/data/perf-exp-1/appliedmanifestworks
>        1       4     142 /Users/mspreitz/data/perf-exp-1/workstatuses
>        1       4      80 /Users/mspreitz/data/perf-exp-1/manifestworks
>        1       7     104 /Users/mspreitz/data/perf-exp-1/deployments-wec
>       10      30     440 /Users/mspreitz/data/perf-exp-1/secrets-wds1
>      106     340    5398 total
> ```
> 
> Some example contents (with .txt appended to the filename):
> 
> [deployments-wds1.txt](https://github.com/user-attachments/files/16948850/deployments-wds1.txt) [deployments-wec.txt](https://github.com/user-attachments/files/16948851/deployments-wec.txt) [services-wds1.txt](https://github.com/user-attachments/files/16948853/services-wds1.txt) [services-wec.txt](https://github.com/user-attachments/files/16948854/services-wec.txt)

@MikeSpreitzer, the contents of these output files looks great - as expected! Thanks for posting them here

dumb0002 commented:
> I am testing the latest version. Commit [754e4a9](https://github.com/kubestellar/kubestellar/commit/754e4a9a67b655a175db06f7cb0fc475cf8d8977).
> 
> I noticed that the end of the output from running clusterlloader2 is
> 
> ```shell
> JUnit report was created: /Users/mspreitz/go/src/k8s.io/perf-tests/clusterloader2/junit.xml
> I0910 13:34:16.850798    5974 prometheus.go:307] Get snapshot from Prometheus
> ```
> 
> And the exit status is 0. Where is the Prometheus snapshot? BTW, as far as I know, my system does not have Prometheus installed.

@MikeSpreitzer, that's just a default log message. Prometheus is not installed indeed. But you can also install the prometheus stack with clusterloader2 by setting the following flag `--enable-prometheus-server=true`

MikeSpreitzer reviewed (approved):
/lgtm
/approve
Thanks!

dumb0002 commented:
> > But you can also install the prometheus stack with clusterloader2 by setting the following flag `--enable-prometheus-server=true`
> 
> But that is not what you are planning to do for your study, right?

Correct, we are installing our own prometheus stack instead as it is described here: https://github.com/dumb0002/kubestellar/blob/collect-creation-time/monitoring/README.md
